[REVIEW] Least Angle Regression

tfeher opened the pull request:
This PR implements Least Angle Regression (LARS). 

Lars is a model selection method, we select a number of features, for the prediction (controlled by the n_nonzero_coefs arg) and determine their regression coefficients. 

The solver is implemented according to the [paper by Efron, Hastie, Johnstone and Tibshirani](http://statweb.stanford.edu/~tibs/ftp/lars.pdf).

This PR depends on RAFT PRs https://github.com/rapidsai/raft/pull/94 and https://github.com/rapidsai/raft/pull/95 and https://github.com/rapidsai/raft/pull/102, and https://github.com/rapidsai/raft/pull/103.

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

teju85 commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,705 @@
+                           MLCommon::device_buffer<math_t>& workspace,
+                           cudaStream_t stream) {
+  const idx_t align_bytes = 16 * sizeof(math_t);
+  // We might need to start a few elements earlier to ensure that the unary
```
we discussed this during our 1:1 today and unary-op does look at the proper alignment of the input addresses. So, this extra alignment shouldn't be needed, right?

teju85 commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,705 @@
+                                        n_rows, n_cols, &one, X, ld_X, y, 1,
+                                        &zero, cor.data(), 1, stream));
+  if (coef_path) {
+    std::cout << "Max iter " << *max_iter << "\n";
```
nitpick: stray print?

teju85 commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,705 @@
+    const int TPB = 64;
+    MLCommon::Cache::get_vecs<<<raft::ceildiv(n_active, TPB), TPB, 0, stream>>>(
+      X, ld_X, active_idx, n_active, X_active_cols.data());
+    CUDA_CHECK(cudaPeekAtLastError());
```
better to use `cudaGetLastError()`

teju85 reviewed (changes requested):
Can you also provide some perf numbers here for reference purposes?

tfeher reviewed (commented):
Thanks @teju85 for the comments, I have addressed the issues.

teju85 commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,717 @@
+  math_t one = 1;
+  math_t zero = 0;
+  // Set initial correlation to X.T * y
+  // TODO: this returns NaNs if n_rows >= 65536 and math_t = fp32
```
Do you have a repro for this? If yes, can you file a bug against cublas for this one?

tfeher replied:
Further investigations showed that it is not a cuBLAS error, but a tricky problem with the input data. I have opened issue #3189 to track this. Additionally, I will disable the fp32 solver path until the issue is resolved.

tfeher commented:
I have moved LARS to the experimental namespace. There two known problems:
- Handling collinear input. I have added a [python unit test](https://github.com/tfeher/cuml/blob/97c45376334a80524f3acdefc488abd411778faf/python/cuml/test/test_lars.py#L102) for that and working on a fix. Scikit learn detects the problem using [this check](https://github.com/scikit-learn/scikit-learn/blob/d640b7fc61ce716af9d113a7c92c953c1ec3e36f/sklearn/linear_model/_least_angle.py#L605) I will evaluate whether this is the best course for us.
- [cuBLAS error](https://github.com/tfeher/cuml/blob/97c45376334a80524f3acdefc488abd411778faf/cpp/src/solver/lars_impl.cuh#L480-L483) in fp32 mode. The [fp32 stress test](https://github.com/tfeher/cuml/blob/97c45376334a80524f3acdefc488abd411778faf/python/cuml/test/test_lars.py#L53) is a reproducer on V100 CUDA 10.2. I will check with other library versions, whether the error exists.

tfeher commented:
@teju85 Here are some preliminary perf numbers, measured (V100 vs  Intel Xeon E5-2698). This was measured with default params for the regressor, using a synthetic dataset generated by make_regression. Note that LARS usually works with ncol x ncol Gram matrix therefore sckit-learn's solver is very fast for a large fraction of the parameter space. Our GPU solver improves on that and ensures that we remain fast even with a large number of rows or columns.
![lars_cpp_speedup](https://user-images.githubusercontent.com/3671106/100101675-d6c77280-2e62-11eb-821a-f17ff2930430.png)

tfeher commented:
Removed the waiting on author label since the two [remaining issues](https://github.com/rapidsai/cuml/pull/3160#issuecomment-732189602) that I am addressing are not affecting the Cython wrappers.

drobison00 commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+    handle.get_cublas_handle(), CUBLAS_SIDE_LEFT, fillmode, CUBLAS_OP_T,
+    CUBLAS_DIAG_NON_UNIT, n_active, 1, &alpha, U, ld_U, ws, ld_U, stream));
+
+  // w0 stores x, the solution of U.T x = sign_A. Now we solve U * ws = x
```
```suggestion
  // ws stores x, the solution of U.T x = sign_A. Now we solve U * ws = x
```

tfeher replied:
Thanks for catching this!

drobison00 reviewed (approved):
Went over the cuh and pyx implementation code, looks good to me. Noted a couple typos / missing params.

tfeher commented on python/cuml/linear_model/lars.pyx:
```diff
@@ -0,0 +1,358 @@
+        Return the Gram matrix, or None if it is not applicable.
+        """
+        Gram = None
+        # TODO check once more the n_cols / n_rows condition
```
Check updated and TODO removed.

tfeher commented on python/cuml/linear_model/lars.pyx:
```diff
@@ -0,0 +1,358 @@
+        if self.eps is None:
+            self.eps = np.finfo(np.float).eps
+        else:
+            # TODO propagate eps to the solver
```
Eps is now propageted to the solver.

tfeher commented on python/cuml/linear_model/lars.pyx:
```diff
@@ -0,0 +1,358 @@
+            Gram_ptr = Gram_m.ptr
+        cdef uintptr_t coef_path_ptr = <uintptr_t> nullptr
+        if (self.fit_path):
+            # TODO catch memory error and give informative message if max_iter
```
Memory error handling was improved in commit cf9b5d6, todo removed.

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+            cudaStream_t stream) {
+  const cublasFillMode_t fillmode = CUBLAS_FILL_MODE_UPPER;
+
+  // First we alculate x by solving equation U.T x = sign_A.
```
corrected

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+ * @param X device array  of feature vectors in column major format, size
+ *     [n_rows * n_cols]
+ * @praam n_rows number of training vectors
+ * @param n_cols number of features
```
added

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+ * @param n_cols number of features
+ * @parma sign array with sign of the active set, size [n_cols]
+ * @param U device pointer to the Cholesky decomposition of G0,
+ *     size [n_cols * n_cols]
```
added

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+ *     size [n_cols * n_cols]
+ * @param G0 device pointer to Gram matrix G0 = X.T*X (can be nullptr),
+ *     size [n_cols * n_cols]. Note the difference between G0 and
+ *     GA = X_A.T * X_A
```
added

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+ *     size [n_cols * n_cols]. Note the difference between G0 and
+ *     GA = X_A.T * X_A
+ * @param workspace workspace for the Cholesky update
+ * @param w device pointer, size [n_active]
```
corrected

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+ *  http://statweb.stanford.edu/~tibs/ftp/lars.pdf
+ *
+ * @param handle RAFT handle
+ * @param max_iter maximum number of iterations
```
added

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+ * @param A device pointer to a scalar, as defined by eq 2.5 in [1]
+ * @param cor device pointer to correlation vector, size [n_active]
+ * @param G device pointer to Gram matrix of the active set (without signs)
+ *    size [n_active * ld_G]
```
added

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+ * @param u device pointer to equiangular vector size [n_rows]. Only used if the
+ *     Gram matrix G is not available.
+ * @param ws device pointer to the ws vector defined in calcEquiangularVec,
+ *    size [n_active]
```
added

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+}
+
+/**
+ * @brief Initialize for Lars training.
```
No, added docstring.

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+
+/**
+ * @brief Update regression coefficient and correlations
+ */
```
No, corrected the sloppiness.

tfeher commented on cpp/src/solver/lars_impl.cuh:
```diff
@@ -0,0 +1,747 @@
+ * @param G device pointer to Gram matrix of the active set (without signs)
+ *    size [n_active * ld_G]
+ * @param X device array of training vectors in column major format,
+ *     size [n_rows * n_cols]. Only used if the gram matrix is not avaiable.
```
added

tfeher reviewed (commented):
Thanks @drobison00 for the review, I have addressed the issues.

Additionally, collinearity detection was added to the cpp solver, and the related pytest are now enobled.

JohnZed commented:
Because this is far along in review and going into the experimental namespace, I believe we can still push it for 0.17 (assuming no new issues emerge) despite being in burndown now.

tfeher commented:
I believe the solver is fairly robust in fp64. The bug in fp32 mode (#3189) is avoided by automatically converting the data to fp64. I am not aware of any other issues. Will need to merge https://github.com/rapidsai/raft/pull/94 before the CI can build this.

dantegd commented:
@tfeher rapidsai/raft#94 has been merged

cjnolet commented:
rerun tests

tfeher commented:
The cpp unit tests are incorrectly marked as passed (although the error is unrelated to LARS, the problem is with columnSort #3196).

JohnZed commented on python/cuml/test/test_lars.py:
```diff
@@ -0,0 +1,191 @@
+@pytest.mark.parametrize(
+    "column_info",
+    [
+        unit_param([20, 10]),
```
would be nice to try the case of 1 column as well (future addition)

JohnZed commented on python/cuml/test/test_lars.py:
```diff
@@ -0,0 +1,191 @@
+        sklars.fit(X_train, y_train)
+
+        assert cu_score_train >= sklars.score(X_train, y_train) - 0.05
+        assert cu_score_test >= sklars.score(X_test, y_test) - 0.1
```
This gap of 0.1 seems pretty large? How big is it in relative terms? is this going from 0.95 to 0.85? Can we narrow it down?

tfeher replied:
You are right, this is very broad. Actually the accuracy is 1.0 for these make_regression datasets without noise. I have narrowed down the tolerance to 0.1% absolute accuracy diff.

JohnZed commented on python/cuml/test/test_lars.py:
```diff
@@ -0,0 +1,191 @@
+                                    {"n_nonzero_coefs": 5},
+                                    {"n_nonzero_coefs": 2}])
+def test_lars_attributes(datatype, params):
+    X, y = load_boston(return_X_y=True)
```
Maybe try one variant with `fit_intercept=False` (some of that codepath seems to be uncovered)

tfeher replied:
Yes, I was already working on this yesterday, just got sidelined with other issues. Added the test.

After this only the memory errors handling is not covered. I will open a separate issue for that to be tasted later.

JohnZed reviewed (approved):
Looks good! I have a few smaller suggestions (mostly little tests and a question). For an experimental module like this, I believe these can be handled in follow-ons, so I'm pre-approving.

tfeher reviewed (commented):
Thanks @JohnZed  for the review, I have fixed the issues!
